### PR TITLE
Error on unsupported batch norm third derivatives

### LINKED ISSUE #186256: `native_batch_norm` silently produces incorrect 3rd+ order derivatives in training mode with `create_graph=True`
### 🐛 Describe the bug

Repeated differentiation (`create_graph=True`) through the fused `native_batch_norm` in **training mode** silently produces **incorrect derivatives from 3rd order onward**. Orders 0–2 are correct; order ≥3 diverges from the mathematically correct value by large amounts (not floating-point noise). Because no error is raised, this silently corrupts any computation relying on 3rd+ order derivatives of batch norm.

This is the same root cause as #175754 for `native_layer_norm`: the saved `mean`/`invstd` tensors handed to the fused double-backward have no autograd history connecting them to the input, so when autograd differentiates *through* the double backward for 3rd+ order derivatives it treats them as constants.

It only affects **training mode** (where `mean`/`invstd` depend on the input); eval mode is an affine map with frozen statistics and is correct to all orders.

### To reproduce

Compare the fused op against a mathematically identical decomposition from primitives, taking directional derivatives `dᵏ/dtᵏ ⟨f(x + t·v), w⟩` at `t = 0` via repeated `autograd.grad(..., create_graph=True)`:

```python
import torch
from torch.nn.functional import batch_norm

torch.manual_seed(0)
torch.set_default_dtype(torch.float64)

N, C = 8, 3
x = torch.randn(N, C)
v = torch.randn(N, C)             # direction
w = torch.randn(N, C)            # cotangent -> scalar output

running_mean, running_var = torch.zeros(C), torch.ones(C)

def fused(z):                    # fused native_batch_norm (training)
    return batch_norm(z, running_mean, running_var, None, None,
                      training=True, momentum=0.0, eps=1e-5)

def manual(z):                   # mathematically identical, from primitives
    mean = z.mean(0, keepdim=True)
    c = z - mean
    var = (c * c).mean(0, keepdim=True)
    return c * (var + 1e-5).rsqrt()

def t_derivatives(f, order):     # d^k/dt^k <f(x + t v), w> at t = 0
    t = torch.zeros((), requires_grad=True)
    g = (f(x + t * v) * w).sum()
    coeffs = []
    for k in range(order + 1):
        coeffs.append(g.detach().clone())
        if k < order:
            (g,) = torch.autograd.grad(g, t, create_graph=True)
    return coeffs

fused_d = t_derivatives(fused, 5)
manual_d = t_derivatives(manual, 5)
print("order | match | max_abs_diff")
for k in range(6):
    match = torch.allclose(fused_d[k], manual_d[k])
    print(f"  {k}   |  {str(match):5} | {(fused_d[k] - manual_d[k]).abs().max().item():.3e}")
```

### Output (should match everywhere, but does not):
```
order | match | max_abs_diff
  0   |  True  | 0.000e+00
  1   |  True  | 0.000e+00
  2   |  True  | 8.882e-16
  3   |  False | 1.258e+00
  4   |  False | 7.850e+01
  5   |  False | 2.695e+02
```

### Expected behaviour

Either the fused training-mode double-backward should produce correct 3rd+ order derivatives (matching the primitive decomposition), or — as a strict improvement until then — it should raise rather than return silently-wrong values (cf. the layer_norm proposal in #176234).

### Related

- #175754 — same bug for `native_layer_norm` (closed; root cause documented there)
- #176234 — proposed layer_norm guard; reverted, not on `main`

### Versions

```
Collecting environment information...
PyTorch version: 2.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.5.1 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.1.1.101)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.6 | packaged by Anaconda, Inc. | (main, Oct  3 2024, 02:26:31) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-26.5.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M5

Versions of relevant libraries:
[pip3] backpack-for-pytorch==1.7.1
[pip3] jet-for-pytorch==0.0.2.dev133+gf171ce318
[pip3] numpy==2.4.6
[pip3] torch==2.12.0
[pip3] torchvision==0.27.0
[conda] Could not collect
```

cc @ezyang @gchanan @kadeng @msaroufim @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

### PR DESCRIPTION
**Pull request by he-yufeng**
Fixes #186256.

`batchnorm_double_backward` analytically accounts for the dependence of the saved training statistics on the input, so second-order gradients are correct. Those saved tensors do not carry autograd history, though, and differentiating through the manual double backward silently treats them as constants.

This change attaches a delayed error to the training statistics when a further derivative is requested. Evaluation mode is unchanged because its running statistics are constants. The regression test is skipped in compiled autograd, which does not support the higher-order gradient sequence required by the test; this avoids the CI failure mode that reverted the analogous layer-norm guard in #176234.

## Validation

- Reproduced the current behavior with installed `torch==2.12.0+cpu`: a third-order batch-norm gradient is silently computed.
- `python -m py_compile test/test_autograd.py test/inductor/test_compiled_autograd.py`
- `git diff --check`

## To verify

- Source build and focused `test_autograd.py -k batch_norm_errors_on_third_order_grad`
- Compiled-autograd test collection confirms the focused test is skipped
- Full PyTorch CI

Local source-build and `lintrunner -a` validation were not available in this Windows worktree: it has no generated `torch.version`, and `spin` / `lintrunner` are not installed.

AI assistance was used to inspect the existing layer-norm precedent and prepare the patch; the duplicate search, implementation, and validation boundaries were reviewed against current main.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -4865,17 +4865,28 @@ std::tuple<Tensor, Tensor, Tensor> batchnorm_double_backward(
...
+  Tensor invstd;
+  if (training) {
+    mean = toNonOptTensor(save_mean).to(input.scalar_type());
+    invstd = toNonOptTensor(save_invstd).to(input.scalar_type());
+    if (at::GradMode::is_enabled() && input.requires_grad()) {
+      auto node = c10::make_intrusive<DelayedError>(
+          "batch_norm does not support 3rd+ order derivatives.",
+          /* num inputs */ 3);
+      auto result = node->apply({mean, invstd, input});
+      mean = result[0];
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp]: Can these be std::moved out of result?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -4865,17 +4865,28 @@ std::tuple<Tensor, Tensor, Tensor> batchnorm_double_backward(
...
+      invstd = result[1];
+    }
+  } else {
+    mean = toNonOptTensor(running_mean);
+    invstd = toNonOptTensor(running_var).add(Scalar(eps)).pow(-0.5);
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp]: Can this be inplace pow_? Ideally add_ but I am not sure that works with a scalar arg. Or it might need an isTensorSubclass check for NestedTensor etc and not worth it I guess.

**Comment by he-yufeng** [comment]
Addressed both review nits in the latest push:

- moved `result[0]` / `result[1]` out of the DelayedError result vector before assigning them back to `mean` and `invstd`;
- changed the inference `pow(-0.5)` to `pow_(-0.5)` on the temporary tensor returned by `add(...)`.

I kept `add(...)` out-of-place rather than `add_(...)` so the running variance tensor is not modified while forming `invstd`.

Local check: `git diff --check` on the updated patch.

**Comment by he-yufeng** [comment]
@Skylion007 both nits are addressed — `result[0]`/`result[1]` are moved out with `std::move`, and the inference path now uses the in-place `pow_`. CI has been green since. Could you re-review when you have a moment?

**Comment by soulitzer** [comment]
@claude review these changes

**Comment by he-yufeng** [comment]
Thanks for the review @soulitzer. Added the explanatory comment in 9efbe0ae3b — it spells out why `input` is in the `node->apply({mean, invstd, input})` call: `save_mean`/`save_invstd` don't require grad, so without a grad-requiring input the node would be marked non-executable and the `Error` grad_fn would never be installed (the guard would silently no-op). `result[2]` stays unused.

Kept the test CPU-only since it exercises an error path / control flow rather than numerics. It's a comment-only change, so behavior is unchanged.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related to this PR?

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by albanD** [comment]
## mergedog handoff

All CI is green (or skipped). Ready for human review and `@pytorchbot merge`.


Current PR head: `cb4b86114e66a444fd62fe1d1177608e828a5357`.
No mergedog-authored commits were pushed during this run.

No LLM was invoked during this run.

**Comment by albanD** [comment]
## mergedog handoff (after merge failure)

pytorchmergebot reported `Merge failed`. mergedog re-inspected CI and is handing back off to you. **Please review the latest session(s) below and re-run `@pytorchbot merge` if you're happy** — mergedog will not retrigger the merge itself.


Current PR head: `cb4b86114e66a444fd62fe1d1177608e828a5357`.
No mergedog-authored commits were pushed during this run.

No LLM was invoked during this run.

**Comment by albanD** [comment]
## mergedog handoff (after merge failure)

pytorchmergebot reported `Merge failed`. mergedog re-inspected CI and is handing back off to you. **Please review the latest session(s) below and re-run `@pytorchbot merge` if you're happy** — mergedog will not retrigger the merge itself.


Current PR head: `cb4b86114e66a444fd62fe1d1177608e828a5357`.
No mergedog-authored commits were pushed during this run.

No LLM was invoked during this run.

**Comment by albanD** [comment]
## mergedog handoff (after merge failure)

pytorchmergebot reported `Merge failed`. mergedog re-inspected CI and is handing back off to you. **Please review the latest session(s) below and re-run `@pytorchbot merge` if you're happy** — mergedog will not retrigger the merge itself.


Current PR head: `cb4b86114e66a444fd62fe1d1177608e828a5357`.
No mergedog-authored commits were pushed during this run.

No LLM was invoked during this run.
